android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java

```diff
@@ -1289,11 +1289,13 @@ public void progress(int progress) {
 				}
 
 			} catch(Throwable t) {
-				if (client != null) {
-					Log.d(TAG, "clearing the expired and idle connections", Log.DEBUG_MODE);
-					client.disconnect();
-				} else {
-					Log.d(TAG, "client is not valid, unable to clear expired and idle connections");
+				synchronized (client) {
```

Comment:
if client *is* null, doesn't the synchronized block break?

Reply:
Chis is right. From the looks of it, calling abort() during a long download should cause a NullPointerException.

Also, we're not synchronizing against the "client" object anywhere else in the code, making this synchronization block useless.

The race condition that can happen here is with the `abort()` method since it too is accessing and assigning the `client` member variable. This is the only spot in the code where the `client` member variable is accessed on a different thread. While we could wrap a synchronization block around all `client` assignments, in this case, it might be simpler to ditch the synchronization block and do the following in the `abort()` method instead...

```java
public void abort()
{
	if (readyState > READY_STATE_UNSENT && readyState < READY_STATE_DONE) {
		aborted = true;
//<MyNewCode>
		try {
			HttpURLConnection client = this.client;
			if (client != null) {
				client.disconnect();
			}
		} catch (Exception ex) {
		}
//</MyNewCode>
		// Fire the disposehandle event if the request is aborted.
		// And it will dispose the handle of the httpclient in the JS.
		proxy.fireEvent(TiC.EVENT_DISPOSE_HANDLE, null);
	}
}
```

The idea is that we don't want to set `client` to null in the `abort()` method. We should let the Runnable thread do it in its `finally` block (it's already has `abort` member variable checks). Also, Java reference assignment is atomic and copying the `client` member variable to a local like the above avoids a race condition where the `client` member variable reference might change on the other thread, such as to null or to another client connection when redirecting.

Reply:
Good catch! I'll update the PR